### PR TITLE
Message Passing Neural Network

### PR DESCRIPTION
**Pull request by miaecle**
This PR adds in:

- New model: Message passing neural network(with the default edge network and set2set readout function)
details in https://arxiv.org/abs/1511.06391
This model is pretty similar to Weave, they can share the same featurizer.(example in `delaney_MPNN.py`) Another option is added to WeaveFeaturizer to replace graph distance by Euclidean distance(designed only for QM8 and QM9, example in `qm8_MPNN.py`)

- Add checking explicit hydrogen option to `WeaveFeaturizer` and `ConvMolFeaturizer`

- A minor modification in `dataset.iterbatches` function, @rbharath could you help me check if this will cause any serious breaks?

#600 #507 
Also, credit to @patrickhop for the implementation of pytorch version of MPNN, it is very helpful and instructive!

### CODE REVIEW

**Comment by patrickhop** [comment]
Awesome work @miaecle. Looking forward to giving it a try.

File: deepchem/feat/graph_features.py [hunk 1]
```diff
@@ -111,83 +111,100 @@ def atom_to_id(atom):
...
+        one_of_k_encoding_unk(atom.GetImplicitValence(), [0, 1, 2, 3, 4, 5, 6]) + \
+        [atom.GetFormalCharge(), atom.GetNumRadicalElectrons()] + \
         one_of_k_encoding_unk(atom.GetHybridization(), [
             Chem.rdchem.HybridizationType.SP, Chem.rdchem.HybridizationType.SP2,
             Chem.rdchem.HybridizationType.SP3, Chem.rdchem.HybridizationType.
             SP3D, Chem.rdchem.HybridizationType.SP3D2
-        ]) + [atom.GetIsAromatic()])
+        ]) + [atom.GetIsAromatic()]
+    else:
+      results = results + one_of_k_encoding_unk(atom.GetTotalNumHs(), [0, 1, 2, 3, 4]) + \
```
Comment (lilleswing) [inline deepchem/feat/graph_features.py]: can you just append 
```python 
one_of_k_encoding_unk(atom.GetTotalNumHS(), [0,1,2,3,4]) 
```
at the end of the feature vector if explicit_H instead of duplicating the code in both branches?

``` python
results = results + [atom.GetFormalCharge(), atom.GetNumRadicalElectrons()] + \
          one_of_k_encoding_unk(atom.GetHybridization(), [
          Chem.rdchem.HybridizationType.SP, Chem.rdchem.HybridizationType.SP2,
          Chem.rdchem.HybridizationType.SP3, Chem.rdchem.HybridizationType.
          SP3D, Chem.rdchem.HybridizationType.SP3D2
          ]) + [atom.GetIsAromatic()] + one_of_k_encoding_unk(atom.GetTotalNumHS(), [0,1,2,3,4]) 
if explicit_H:
  results = results + one_of_k_encoding_unk(atom.GetImplicitValence(), [0, 1, 2, 3, 4, 5, 6])
```

File: deepchem/feat/graph_features.py [hunk 2]
```diff
@@ -263,14 +290,30 @@ class WeaveFeaturizer(Featurizer):
 
   name = ['weave_mol']
 
-  def __init__(self):
+  def __init__(self, graph_distance=True, explicit_H=None):
```
Comment (lilleswing) [inline deepchem/feat/graph_features.py]: why not default explicit_H to False instead of None?

File: deepchem/feat/graph_features.py [hunk 3]
```diff
@@ -263,14 +290,30 @@ class WeaveFeaturizer(Featurizer):
...
+      # Set to True if explicit hydrogen is not specified
+      self.check_H = True
 
   def _featurize(self, mol):
     """Encodes mol as a WeaveMol object."""
+    # Check hydrogen in the molecule
+    if self.check_H and not self.explicit_H:
+      for a in mol.GetAtoms():
+        if a.GetSymbol() == 'H':
+          self.explicit_H = True
```
Comment (lilleswing) [inline deepchem/feat/graph_features.py]: can we get into a situation where we are featurizing many molecules and the first molecule does not have explicit H and later ones do?
Right now we will break in weird ways, because we are lazily setting this.

File: deepchem/feat/graph_features.py (hunk 2 above)
Comment (miaecle) [inline deepchem/feat/graph_features.py]: It's like if you call this function with explicit_H=False, then we will not check hydrogens. But if you do not specify, we will check by default.

File: deepchem/feat/graph_features.py (hunk 3 above)
Comment (miaecle) [inline deepchem/feat/graph_features.py]: I actually don't know, since rarely organic molecules do not have hydrogens, maybe we could just get rid of this check hydrogen part and specify with or without hydrogen upon calling.

File: deepchem/models/tensorgraph/graph_layers.py [hunk 4]
```diff
@@ -799,3 +799,209 @@ def DAGgraph_step(self, batch_inputs, W_list, b_list):
       outputs = tf.nn.xw_plus_b(outputs, W, b_list[idw])
       outputs = self.activation(outputs)
     return outputs
+
+
+class MessagePassing(Layer):
```
Comment (miaecle) [inline deepchem/models/tensorgraph/graph_layers.py]: For sure, will do!

File: deepchem/data/datasets.py [hunk 5]
```diff
@@ -96,6 +96,8 @@ def pad_batch(batch_size, X_b, y_b, w_b, ids_b):
 
     # Fill in batch arrays
     start = 0
+    # Only the first set of copy will be counted in training loss
```
Comment (lilleswing) [inline deepchem/data/datasets.py]: nit: you can completely remove the else on line 83

File: deepchem/data/datasets.py [hunk 6]
```diff
@@ -96,6 +96,8 @@ def pad_batch(batch_size, X_b, y_b, w_b, ids_b):
...
+    w_out[start:start + num_samples] = w_b[:]
```
Comment (lilleswing) [inline deepchem/data/datasets.py]: the real question is why can't we use this technique for X_out and y_out as well

File: deepchem/feat/graph_features.py (hunk 3 above)
Comment (lilleswing) [inline deepchem/feat/graph_features.py]: Yeah I would rather set it in the constructor and throw an exception if our assumptions from the constructor aren't met.

**Comment by rbharath** [comment]
Thanks for the PR! Having MPNNs inside DeepChem will be very useful.

The comments on this thread look pretty solid. I think we can merge once those issues are fixed.

File: deepchem/data/datasets.py (hunk 5 above)
Comment (miaecle) [inline deepchem/data/datasets.py]: Already modified!

File: deepchem/data/datasets.py (hunk 6 above)
Comment (miaecle) [inline deepchem/data/datasets.py]: I think basically this iterbatches function should work for all kinds of dataset structures(different featurizers generate different X_b), which makes it hard to have a uniform way of padding. The change on w_b should be enough for the loss to be correctly mapped, so I am personally satisfied with the structure now.

File: deepchem/feat/graph_features.py (hunk 1 above)
Comment (miaecle) [inline deepchem/feat/graph_features.py]: Got it fixed!

File: deepchem/feat/graph_features.py (hunk 2 above)
Comment (miaecle) [inline deepchem/feat/graph_features.py]: I just deleted everything tricky here, so it should be directly depending on the input parameter `explicit_H` now.

File: deepchem/models/tensorgraph/graph_layers.py (hunk 4 above)
Comment (miaecle) [inline deepchem/models/tensorgraph/graph_layers.py]: Already added for all graph layers, also unit tests in `test_layers_pickle.py`

**Comment by miaecle** [comment]
@lilleswing I have done some refinements, would you mind taking a second look? Thanks for your suggestions!

File: deepchem/data/datasets.py [hunk 7]
```diff
@@ -80,34 +80,35 @@ def pad_batch(batch_size, X_b, y_b, w_b, ids_b):
...
+
+  num_tasks = y_b.shape[1]
+  y_out = np.zeros((batch_size, num_tasks), dtype=y_b.dtype)
+  w_out = np.zeros((batch_size, num_tasks), dtype=w_b.dtype)
+  ids_out = np.zeros((batch_size,), dtype=ids_b.dtype)
+
+  # Fill in batch arrays
+  start = 0
+  # Only the first set of copy will be counted in training loss
+  w_out[start:start + num_samples] = w_b[:]
```
Comment (rbharath) [inline deepchem/data/datasets.py]: I think you can do this same trick for `X_out`, `y_out`, and `ids_out` and remove the for-loop below without effect.

File: deepchem/models/tensorgraph/graph_layers.py [hunk 8]
```diff
@@ -799,3 +841,258 @@ def DAGgraph_step(self, batch_inputs, W_list, b_list):
...
+  """ General class for MPNN
+  default structures built according to https://arxiv.org/abs/1511.06391 """
+
+  def __init__(self,
+               T,
+               message_fn='enn',
+               update_fn='gru',
+               n_hidden=100,
+               **kwargs):
+    """
```
Comment (rbharath) [inline deepchem/models/tensorgraph/graph_layers.py]: The indentation here looks a little weird. Maybe try updating `yapf` to the latest version to see if the artifact vanishes.

File: deepchem/models/tensorgraph/graph_layers.py [hunk 9]
```diff
@@ -799,3 +841,258 @@ def DAGgraph_step(self, batch_inputs, W_list, b_list):
...
+    self.A = None,
+    trainable_weights = self.trainable_weights
+    self.trainable_weights = []
+    return A, trainable_weights
+
+  def set_tensors(self, tensor):
+    self.A, self.trainable_weights = tensor
+
+
+class GatedRecurrentUnit(object):
```
Comment (rbharath) [inline deepchem/models/tensorgraph/graph_layers.py]: Is this the same as https://github.com/deepchem/deepchem/blob/master/deepchem/models/tensorgraph/layers.py#L401, the existing GRU class? Could we replace it and use the existing GRU?

File: deepchem/data/datasets.py (hunk 7 above)
Comment (miaecle) [inline deepchem/data/datasets.py]: I have tried that, but basically in the process of building feed_dict, each sample need to be an actual molecule(there are some functions like `get_num_atoms()` will raise errors on zero padding)

File: deepchem/models/tensorgraph/graph_layers.py (hunk 8 above)
Comment (miaecle) [inline deepchem/models/tensorgraph/graph_layers.py]: Thanks for pointing out, got it fixed!

File: deepchem/models/tensorgraph/graph_layers.py (hunk 9 above)
Comment (miaecle) [inline deepchem/models/tensorgraph/graph_layers.py]: Basically they are the same, but regular tensorflow GRU implementation need all your inputs and hidden states for each step to be connected, while here there exist another phase(EdgeNetwork) that will change the hidden state between steps. My implementation is essentially like a GRUStep function, which only pushes one step forward, kind of similar to the `LSTMStep` in `SetGather`

**Comment by lilleswing** [comment]
@miaecle this PR is incredibly exciting!  I'm especially interested in playing around with the set2set layer.
If possible we should try to unite the GRU layers into a single layer -- maybe with flags.
LGTM otherwise

**Comment by rbharath** [comment]
Seconding @lilleswing's comments! This is a great addition to the DeepChem tool suite. I'm good to go ahead and merge these changes in. Let's figure out the GRU questions on a new issue/PR.

LGTM
